### PR TITLE
TINY-7737: Table buttons enable only when whole selection is within the same table

### PR DESCRIPTION
**Pull request by HAFRMO**
Description of Changes:
Now check that the start and end are both within the same table.

Pre-checks:
* [x] Changelog entry added
* [x] Tests have been added (if applicable)
* [x] Branch prefixed with `feature/` for new features (if applicable)

Review:
* [x] Milestone set
* [x] Review comments resolved

GitHub issues (if applicable):

### CODE REVIEW

**Comment by HAFRMO** [comment]
Tests to be added.

File: modules/tinymce/src/plugins/table/test/ts/browser/settings/TableToolbarTest.ts [hunk 1]
```diff
@@ -50,4 +52,55 @@ describe('browser.tinymce.plugins.table.TableToolbarTest', () => {
...
+
+  context('Ensure table buttons are enabled when appropriate', () => {
+    it('TINY-7737: When the whole selection is in the table', async () => {
+      const editor = await pCreateEditor('tablecaption');
+      editor.focus();
+      editor.setContent(tableWithAdditionalHtml);
+      await TinyUiActions.pWaitForUi(editor, '.tox-tbtn--disabled[title="Table caption"]');
+      TinySelections.setSelection(editor, [ 1, 0, 0, 0 ], 0, [ 1, 0, 0, 0 ], 1);
+      // Wait for a while to allow the toolbar a chance to render
+      await Waiter.pWait(100);
```
Comment (lnewson) [inline modules/tinymce/src/plugins/table/test/ts/browser/settings/TableToolbarTest.ts:64]: This wait should not be here and in the other tests, as the toolbar will always be rendered.

File: modules/tinymce/src/plugins/table/test/ts/browser/settings/TableToolbarTest.ts [hunk 2]
```diff
@@ -50,4 +52,55 @@ describe('browser.tinymce.plugins.table.TableToolbarTest', () => {
     await TinyUiActions.pWaitForUi(editor, '.tox-pop .tox-tbtn[title="Paste row before"]:not(.tox-tbtn--disabled)');
     McEditor.remove(editor);
   });
+
+  context('Ensure table buttons are enabled when appropriate', () => {
```
Comment (lnewson) [inline modules/tinymce/src/plugins/table/test/ts/browser/settings/TableToolbarTest.ts:56]: This isn't a great place for this... this test is checking the context toolbar rendering logic and the `table_toolbar` setting. Your new tests don't use any of that, so please move them to a separate file. You also don't need to be initializing an editor per test, as you aren't changing any settings.

File: modules/tinymce/src/plugins/table/test/ts/browser/settings/ToolbarButtonDisplayTest.ts [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+describe('browser.tinymce.plugins.table.ToolbarButtonDisplayTest', () => {
+  const hook = TinyHooks.bddSetupLight<Editor>({
+    plugins: 'table',
+    toolbar: 'tablecaption',
+    base_url: '/project/tinymce/js/tinymce'
+  }, [ Plugin, Theme ], true);
+
+  const editorContent = '<p>A</p><table><tbody><tr><td>B</td></tr></tbody></table><p>C</p>';
+
+  const displayTest = async (startPath: number[], startOffset: number, endPath: number[], endOffset: number, shouldBeEnabled: boolean) => {
```
Comment (lnewson) [inline modules/tinymce/src/plugins/table/test/ts/browser/settings/ToolbarButtonDisplayTest.ts:19]: This should start with `p` as it's returning a promise. e.g. `pDisplayTest`

File: modules/tinymce/src/plugins/table/test/ts/browser/settings/ToolbarButtonDisplayTest.ts [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+
+    if (shouldBeEnabled) {
+      await TinyUiActions.pWaitForUi(editor, '.tox-tbtn[title="Table caption"]');
+      UiFinder.notExists(SugarBody.body(), '.tox-tbtn--disabled[title="Table caption"]');
+    } else {
+      await TinyUiActions.pWaitForUi(editor, '.tox-tbtn--disabled[title="Table caption"]');
+    }
+  };
+
+  context('Ensure table buttons are enabled when appropriate', () => {
```
Comment (lnewson) [inline modules/tinymce/src/plugins/table/test/ts/browser/settings/ToolbarButtonDisplayTest.ts:33]:
```suggestion
  context('Ensure table caption buttons are enabled when appropriate', () => {
```
This is very specifically targeted at the `tablecaption` button.

File: modules/tinymce/src/plugins/table/test/ts/browser/settings/ToolbarButtonDisplayTest.ts [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+    it('TINY-7737: When the whole selection is in the table', async () => {
+      await displayTest([ 1, 0, 0, 0 ], 0, [ 1, 0, 0, 0 ], 1, true);
+    });
```
Comment (lnewson) [inline modules/tinymce/src/plugins/table/test/ts/browser/settings/ToolbarButtonDisplayTest.ts:36]:
```
    it('TINY-7737: When the whole selection is in the table', () => displayTest([ 1, 0, 0, 0 ], 0, [ 1, 0, 0, 0 ], 1, true));
```
```
    it('TINY-7737: When the whole selection is in the table', () =>
      displayTest([ 1, 0, 0, 0 ], 0, [ 1, 0, 0, 0 ], 1, true)
    );
```
This and others could be shortened to something like one of these 2.

File: modules/tinymce/src/plugins/table/test/ts/browser/settings/ToolbarButtonDisplayTest.ts [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+
+  const displayTest = async (startPath: number[], startOffset: number, endPath: number[], endOffset: number, shouldBeEnabled: boolean) => {
+    const editor = hook.editor();
+    editor.setContent(editorContent);
+    await TinyUiActions.pWaitForUi(editor, '.tox-tbtn--disabled[title="Table caption"]');
+    TinySelections.setSelection(editor, startPath, startOffset, endPath, endOffset);
+
+    if (shouldBeEnabled) {
+      await TinyUiActions.pWaitForUi(editor, '.tox-tbtn[title="Table caption"]');
+      UiFinder.notExists(SugarBody.body(), '.tox-tbtn--disabled[title="Table caption"]');
```
Comment (lnewson) [inline modules/tinymce/src/plugins/table/test/ts/browser/settings/ToolbarButtonDisplayTest.ts:27]:
```suggestion
      await TinyUiActions.pWaitForUi(editor, '.tox-tbtn[title="Table caption"]:not(.tox-tbtn-disabled)');
```
This can be shortened to this I believe.

File: modules/tinymce/src/plugins/table/test/ts/browser/settings/ToolbarButtonDisplayTest.ts (hunk 5 above)
Comment (HAFRMO) [reply modules/tinymce/src/plugins/table/test/ts/browser/settings/ToolbarButtonDisplayTest.ts:36]: Not sure what you mean here.

Comment (lnewson) [reply modules/tinymce/src/plugins/table/test/ts/browser/settings/ToolbarButtonDisplayTest.ts:36]: You can just return the result of the function, so we don't need to do an async/await.

Comment (HAFRMO) [reply modules/tinymce/src/plugins/table/test/ts/browser/settings/ToolbarButtonDisplayTest.ts:36]: Oh, you mean so. Yeah, that's true.

File: modules/tinymce/src/plugins/table/main/ts/selection/SelectionTargets.ts [hunk 7]
```diff
@@ -62,18 +62,25 @@ export const getSelectionTargets = (editor: Editor, selections: Selections): Sel
 
   const isCaption = SugarNode.isTag('caption');
   const isDisabledForSelection = (key: keyof ExtractedSelectionDetails) => selectionDetails.forall((details) => !details[key]);
+  const getStart = () => TableSelection.getSelectionStartCellOrCaption(Util.getSelectionStart(editor), Util.getIsRoot(editor));
+  const getEnd = () => TableSelection.getSelectionEndCellOrCaption(Util.getSelectionEnd(editor), Util.getIsRoot(editor));
+
+  const findTargets = (): Optional<RunOperation.CombinedTargets> =>
```
Comment (ashfordneil) [inline modules/tinymce/src/plugins/table/main/ts/selection/SelectionTargets.ts:68]: Could this be simpler with `Optionals.lift2`?

Something like

```typescript
const start = getStart();
const end = getEnd().bind(TableLookup.table);

return Optionals.lift2(start, end, (startCellOrCaption, endTable) => {
 ...
});
```

That many nested binds looks like an anti pattern to me

File: modules/tinymce/src/plugins/table/main/ts/selection/TableSelection.ts [hunk 8]
```diff
@@ -12,22 +12,27 @@ import { Attribute, Compare, SelectorFind, SugarElement, SugarElements, SugarNod
...
 import { ephemera } from './Ephemera';
 
-const getSelectionStartCellFallback = (start: SugarElement<Node>) =>
-  TableLookup.table(start).bind((table) =>
+const getSelectionCellFallback = (element: SugarElement<Node>, takeLast: boolean) =>
+  TableLookup.table(element).bind((table) =>
     TableSelection.retrieve(table, ephemera.firstSelectedSelector)
-  ).fold(() => start, (cells) => cells[0]);
+  ).fold(
+    () => element,
```
Comment (ashfordneil) [inline modules/tinymce/src/plugins/table/main/ts/selection/TableSelection.ts:19]:
```suggestion
    Fun.constant(element),
```

**Review by ashfordneil — APPROVED** [review]

File: modules/tinymce/src/plugins/table/main/ts/selection/SelectionTargets.ts (hunk 7 above)
Comment (lnewson) [reply modules/tinymce/src/plugins/table/main/ts/selection/SelectionTargets.ts:68]: I did try to do this fwiw, but because we need to return an `Optional` inside we'd end up with the type being `Optional<Optional<RunOperation.CombinedTargets>>` so it still ends up fairly messy to handle that.

Comment (ashfordneil) [reply modules/tinymce/src/plugins/table/main/ts/selection/SelectionTargets.ts:68]: `.bind(Fun.identity)` once at the end might not be a bad trade-off

Comment (lnewson) [reply modules/tinymce/src/plugins/table/main/ts/selection/SelectionTargets.ts:68]: There's `Optionals.flatten` for that instead, but it's still not pretty.

Comment (ashfordneil) [reply modules/tinymce/src/plugins/table/main/ts/selection/SelectionTargets.ts:68]: This feels cleaner to me — it separates the "is this in the same table as the end of the selection" out from the other logic with `.filter`
```typescript
  const findTargets = (): Optional<RunOperation.CombinedTargets> =>
    getStart().filter((startCellOrCaption) => {
      const startTableOpt = TableLookup.table(startCellOrCaption);
      const endTableOpt = getEnd().bind(TableLookup.table);
      return Optionals.lift2(startTableOpt, endTableOpt, Compare.eq).getOr(false)
    }).map((startCellOrCaption) => {
      if (isCaption(startCellOrCaption)) {
        return TableTargets.noMenu(startCellOrCaption);
      } else {
        // We know this is okay, because it has to be Some for the previous filter to pass
        const startTable = TableLookup.table(startCellOrCaption).getOrDie();
        return TableTargets.forMenu(selections, startTable, startCellOrCaption);
      }
    });
```

File: modules/tinymce/src/plugins/table/main/ts/selection/TableSelection.ts [hunk 9]
```diff
@@ -6,28 +6,33 @@
...
-import { Arr, Optionals } from '@ephox/katamari';
+import { Arr, Fun, Optionals } from '@ephox/katamari';
 import { TableLookup } from '@ephox/snooker';
 import { Attribute, Compare, SelectorFind, SugarElement, SugarElements, SugarNode } from '@ephox/sugar';
 
 import { ephemera } from './Ephemera';
 
-const getSelectionStartCellFallback = (start: SugarElement<Node>) =>
-  TableLookup.table(start).bind((table) =>
+const getSelectionCellFallback = (element: SugarElement<Node>, takeLast: boolean) =>
```
Comment (lnewson) [inline modules/tinymce/src/plugins/table/main/ts/selection/TableSelection.ts:15]: The `takeLast` argument here is never used as it's value is always `false` and doesn't really make sense either as the cell fallback logic is used to move from a col element to a cell element, so it should always be the first row/cell we want.

- [ ] Revert `takeLast` logic

File: modules/tinymce/src/plugins/table/test/ts/browser/settings/ToolbarButtonDisplayTest.ts [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+      await TinyUiActions.pWaitForUi(editor, '.tox-tbtn--disabled[title="Table caption"]');
+    }
+  };
+
+  context('Ensure table caption buttons are enabled when appropriate', () => {
+    it('TINY-7737: When the whole selection is in the table', async () =>
+      pDisplayTest([ 1, 0, 0, 0 ], 0, [ 1, 0, 0, 0 ], 1, true)
+    );
+
+    it('TINY-7737: When the selection starts outside the table, but ends in the table', async () =>
```
Comment (lnewson) [inline modules/tinymce/src/plugins/table/test/ts/browser/settings/ToolbarButtonDisplayTest.ts:35]: This should not be `async` as there is no `await` called.

- [ ] Fix incorrect async awaits

File: modules/tinymce/src/plugins/table/test/ts/browser/settings/ToolbarButtonDisplayTest.ts [hunk 11]
```diff
@@ -0,0 +1,47 @@
...
+import Theme from 'tinymce/themes/silver/Theme';
+
+describe('browser.tinymce.plugins.table.ToolbarButtonDisplayTest', () => {
+  const hook = TinyHooks.bddSetupLight<Editor>({
+    plugins: 'table',
+    toolbar: 'tablecaption',
+    base_url: '/project/tinymce/js/tinymce'
+  }, [ Plugin, Theme ], true);
+
+  const editorContent = '<p>A</p><table><tbody><tr><td>B</td></tr></tbody></table><p>C</p>';
```
Comment (lnewson) [inline modules/tinymce/src/plugins/table/test/ts/browser/settings/ToolbarButtonDisplayTest.ts:15]: This is never used elsewhere, so maybe we should inline it?

File: modules/tinymce/src/plugins/table/test/ts/browser/settings/TableToolbarTest.ts [hunk 12]
```diff
@@ -18,6 +18,7 @@ describe('browser.tinymce.plugins.table.TableToolbarTest', () => {
 
   const pCreateEditor = (toolbar: string) => McEditor.pFromSettings<Editor>({
     plugins: 'table',
+    toolbar: 'tablecaption',
```
Comment (lnewson) [inline modules/tinymce/src/plugins/table/test/ts/browser/settings/TableToolbarTest.ts:21]: Why was this added here? It looks like this wasn't reverted properly when the test logic was moved.

- [ ] Revert this change please

**Review by lnewson — APPROVED** [review]

File: modules/tinymce/src/plugins/table/main/ts/selection/TableSelection.ts [hunk 13]
```diff
@@ -6,28 +6,33 @@
...
+  TableLookup.table(element).bind((table) =>
     TableSelection.retrieve(table, ephemera.firstSelectedSelector)
-  ).fold(() => start, (cells) => cells[0]);
+  ).fold(
+    Fun.constant(element),
+    (cells) => takeLast ? cells[cells.length - 1] : cells[0]);
```
Comment (lnewson) [inline modules/tinymce/src/plugins/table/main/ts/selection/TableSelection.ts:20]: This formatting is incorrect, as the `);` should be on a newline. However, this shouldn't matter when this is reverted, as it it should go back to a single line.

File: modules/tinymce/src/plugins/table/main/ts/selection/TableSelection.ts [hunk 14]
```diff
@@ -6,28 +6,33 @@
...
+const getSelectionStartCaption = getSelectionFromSelector<HTMLTableCaptionElement>('caption', false);
 
-const getSelectionStartCaption = getSelectionStartFromSelector<HTMLTableCaptionElement>('caption');
+const getSelectionStartCell = getSelectionFromSelector<HTMLTableCellElement>('th,td', false);
 
-const getSelectionStartCell = getSelectionStartFromSelector<HTMLTableCellElement>('th,td');
+const getSelectionStartCellOrCaption = getSelectionFromSelector<HTMLTableCellElement | HTMLTableCaptionElement>('th,td,caption', false);
 
-const getSelectionStartCellOrCaption = getSelectionStartFromSelector<HTMLTableCellElement | HTMLTableCaptionElement>('th,td,caption');
+const getSelectionEndCellOrCaption = getSelectionFromSelector<HTMLTableCellElement | HTMLTableCaptionElement>('th,td,caption', false);
```
Comment (lnewson) [inline modules/tinymce/src/plugins/table/main/ts/selection/TableSelection.ts:35]: The implementation of these 2 are identical. This could be changed to just a single function and renamed to `getSelectionCellOrCaption` or similar.

The same could also be done with the other 2 functions, though honestly this whole file confuses me as it's never doing anything with the actual selection. I'm not sure if you want to go that far, but at the very least please just remove the references to `Start` in this file.

- [ ] Cleanup this logic a little

File: modules/tinymce/src/plugins/table/main/ts/selection/TableSelection.ts (hunk 9 above)
Comment (HAFRMO) [reply modules/tinymce/src/plugins/table/main/ts/selection/TableSelection.ts:15]: You are right, it should have been true at one point. My bad, fixed that.

File: modules/tinymce/src/plugins/table/main/ts/selection/TableSelection.ts (hunk 14 above)
Comment (HAFRMO) [reply modules/tinymce/src/plugins/table/main/ts/selection/TableSelection.ts:35]: The error was a boolean that should have been true, to make it pick the last selected cell instead of the first when using the later function. Fixed that.

File: modules/tinymce/src/plugins/table/main/ts/selection/TableSelection.ts (hunk 13 above)
Comment (HAFRMO) [reply modules/tinymce/src/plugins/table/main/ts/selection/TableSelection.ts:20]: It shouldn't be rewerted, since there should be situations where it's used, now when I've added the correct variables. Not quite sure what the situation is, but it should be there now.

File: modules/tinymce/src/plugins/table/test/ts/browser/settings/ToolbarButtonDisplayTest.ts (hunk 11 above)
Comment (HAFRMO) [reply modules/tinymce/src/plugins/table/test/ts/browser/settings/ToolbarButtonDisplayTest.ts:15]: Sure, if you want.

File: modules/tinymce/src/plugins/table/main/ts/selection/SelectionTargets.ts (hunk 7 above)
Comment (HAFRMO) [reply modules/tinymce/src/plugins/table/main/ts/selection/SelectionTargets.ts:68]: Then we're looking up the table twice, and we do the `getOrDie`. That there needs to be a comment saying it's fine speaks against it being an improvement to me.

File: modules/tinymce/src/plugins/table/main/ts/selection/TableSelection.ts (hunk 9 above)
Comment (lnewson) [reply modules/tinymce/src/plugins/table/main/ts/selection/TableSelection.ts:15]: Did you read what I said above? This logic is **incorrect** so it shouldn't have been fixed to use it and should have been removed. To re-iterate this handles the case where the selection ends up in an invalid location within a `<col>` element (Firefox IIRC was the only browser that did this). So when it gets in that state it should **always** be using the first cell as that's always the closest valid position to use. Using the last cell within a selection is definitely the wrong thing to be doing.

Comment (HAFRMO) [reply modules/tinymce/src/plugins/table/main/ts/selection/TableSelection.ts:15]: I read it, but misunderstood what you referred to. My apologizes.

File: modules/tinymce/src/plugins/table/main/ts/selection/SelectionTargets.ts (hunk 7 above)
Comment (ashfordneil) [reply modules/tinymce/src/plugins/table/main/ts/selection/SelectionTargets.ts:68]: Fair enough, let's ship it as is then
